source/common/router/config_impl.cc

```diff
@@ -643,15 +706,56 @@ RouteEntryImplBase::RouteEntryImplBase(const VirtualHostImpl& vhost,
     }
   }
 
+  if (path_rewrite_policy_.enabled()) {
+    if (!prefix_rewrite_.empty() || route.route().has_regex_rewrite()) {
+      throw EnvoyException(
+          "Specify only one of prefix_rewrite, regex_rewrite or path_rewrite_policy");
+    }
+  }
+
+  if (!prefix_rewrite_.empty()) {
+    if (route.route().has_regex_rewrite() || path_rewrite_policy_.enabled()) {
+      throw EnvoyException(
+          "Specify only one of prefix_rewrite, regex_rewrite or path_rewrite_policy");
+    }
+  }
+
   if (route.route().has_regex_rewrite()) {
-    if (!prefix_rewrite_.empty()) {
-      throw EnvoyException("Cannot specify both prefix_rewrite and regex_rewrite");
+    if (!prefix_rewrite_.empty() || path_rewrite_policy_.enabled()) {
+      throw EnvoyException(
+          "Specify only one of prefix_rewrite, regex_rewrite or path_rewrite_policy");
     }
+  }
+
+  if (route.route().has_regex_rewrite()) {
     auto rewrite_spec = route.route().regex_rewrite();
     regex_rewrite_ = Regex::Utility::parseRegex(rewrite_spec.pattern());
     regex_rewrite_substitution_ = rewrite_spec.substitution();
   }
 
+  // Check if pattern rewrite is enabled that the pattern matching is also enabled.
+  // This is needed to match up variable values.
+  if (path_rewrite_policy_.enabled() &&
```

Comment:
I think we need to come up with a more generic way to do this. As written, it only applies to the specific extensions you've introduced and we shouldn't have to update the router every time new extensions are added.

At first glance I'm ok with the idea that a particular path rewrite extension can only be used with a particular path match extension, I just think we need to be a bit more generic in how we check. Perhaps the rewrite policy (or maybe the factory) should have a method that checks the type name of the config object? That would actually allow us to have match and rewrite extensions that could be mixed and matched (though I don't think that's a requirement).

Reply:
In regards to this and below comment on capture group validations. I don't see a good way to move into either extension in a flexible way. By definition we need to look across them and it's a very specific case that isn't likely to be reused.

By adding type checks on extension names we can only run these validations in the case of these two specific extensions together. If someone were to add a different match/rewrite extension the code would just never validation them so no router changes would be needed.

Open to thoughts on how to do this better...

Reply:
I think it would be simple enough to 
1. require a match policy when there's a rewrite policy (I think the reverse is ok -- a custom matcher without a rewrite). This is already in your logic, mixed in with the name checking.
2. if there is a rewrite policy, pass the match predicate to the rewrite predicate and get a result that indicates if the rewrite predicate is capable of rewriting matches from matcher.

Something like:
```c++
if (path_rewrite_policy_.enabled() {
  if (!path_match_policy_.enabled() {
    throw EnvoyException("rewrite_policy requires match_policy");
  }

  // This does the name check and then the variables check.
  absl::optional<std::string> error = path_rewrite_policy_.predicate()->isCompatibleMatchPolicy(path_match_policy_.predicate());
  if (error) {
    throw EnvoyException(*error);
  }
}
```
(There's may be a better type than an optional string for the error, like absl::Status.)

We could relax the rewrite requires match rule in the future (or now) and make the match policy an optional parameter, in which case your extension would report that the appropriate match policy is required if none is passed.

All this means that someone else could write a rewrite predicate extension that depends on this matcher's behavior. I'm happy to have it documented that we don't recommend doing this.

Reply:
Ok this makes sense to me.

In the general case, I do not think we need to require a match policy if we have a rewrite policy.

However I'll add the isCompatibleMatchPolicy validation